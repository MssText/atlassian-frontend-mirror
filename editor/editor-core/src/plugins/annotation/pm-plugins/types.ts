import { PortalProviderAPI } from '../../../ui/PortalProvider';
import { Dispatch, EventDispatcher } from '../../../event-dispatcher';
import { AnnotationTypeProviders } from '../types';

export interface InlineCommentPluginOptions {
  dispatch: Dispatch;
  eventDispatcher: EventDispatcher;
  portalProviderAPI: PortalProviderAPI;
  inlineCommentProvider: AnnotationTypeProviders['inlineComment'];
  pollingInterval?: number;
}
