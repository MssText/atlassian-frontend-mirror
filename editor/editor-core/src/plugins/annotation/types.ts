import React from 'react';
import { AnnotationType } from '@atlaskit/adf-schema';
import { ACTIONS } from './pm-plugins/actions';
import { DecorationSet } from 'prosemirror-view';
import { EditorState, SelectionBookmark } from 'prosemirror-state';

export type AnnotationInfo = {
  id: string;
  type: AnnotationType;
};

export enum AnnotationTypes {
  INLINE_COMMENT = 'inlineComment',
}

type AnnotationComponentProps = {
  /**
   * Selected text (can be used when creating a comment)
   */
  textSelection?: string;

  /**
   * DOM element around selected text (for positioning)
   */
  dom?: HTMLElement;
};

export type AnnotationCreateComponentProps = AnnotationComponentProps & {
  /**
   * Creates an annotation mark in the document with the given id.
   */
  onCreate: (id: string) => void;

  /**
   * Indicates that a draft comment was discarded/cancelled
   */
  onClose?: () => void;
};

export type AnnotationViewComponentProps = AnnotationComponentProps & {
  /**
   * Existing annotations where the cursor is placed.
   * These are provided in order, inner-most first.
   */
  annotations: Array<AnnotationInfo>;

  /**
   * Resolves an annotation with the given ID around the selection.
   */
  onResolve: (id: string) => void;

  /**
   * Removes the annotation from the document
   */
  onDelete?: (id: string) => void;
};

export interface AnnotationState<Type, State> {
  annotationType: Type;
  id: string;
  state: State;
}

export interface AnnotationTypeProvider<Type, State> {
  getState: (
    annotationIds: string[],
  ) => Promise<AnnotationState<Type, State>[]>;
  pollingInterval?: number;
}

export type InlineCommentState = { resolved: boolean };
export type AnnotationTypeProviders = {
  inlineComment: AnnotationTypeProvider<
    AnnotationTypes.INLINE_COMMENT,
    InlineCommentState
  >;
};

export interface AnnotationProvider {
  providers?: AnnotationTypeProviders;
  createComponent?: React.ComponentType<AnnotationCreateComponentProps>;
  viewComponent?: React.ComponentType<AnnotationViewComponentProps>;
}

export type InlineCommentMap = { [key: string]: boolean };
export type InlineCommentAction =
  | {
      type: ACTIONS.INLINE_COMMENT_RESOLVE;
      data: { id: string };
    }
  | {
      type: ACTIONS.SET_INLINE_COMMENT_STATE;
      data: InlineCommentMap;
    }
  | {
      type: ACTIONS.SET_INLINE_COMMENT_DRAFT_STATE;
      data: {
        drafting: boolean;
        editorState: EditorState;
      };
    }
  | {
      type: ACTIONS.INLINE_COMMENT_UPDATE_MOUSE_STATE;
      data: {
        mouseData: {
          x?: number;
          y?: number;
          isSelecting?: boolean;
        };
      };
    };

export type InlineCommentPluginState = {
  annotations: InlineCommentMap;
  mouseData: {
    x: number;
    y: number;
    isSelecting: boolean;
  };
  draftDecorationSet?: DecorationSet;
  bookmark?: SelectionBookmark<any>;
};

const prefix = 'ak-editor-annotation';
export const AnnotationTestIds = {
  prefix,
  floatingComponent: `${prefix}-floating-component`,
  floatingToolbarCreateButton: `${prefix}-toolbar-create-button`,
};

export const DraftDecorationClassName = `${prefix}-draft-selection`;
