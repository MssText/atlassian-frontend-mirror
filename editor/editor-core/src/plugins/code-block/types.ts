export interface CodeBlockOptions {
  allowCopyToClipboard?: boolean;
}
