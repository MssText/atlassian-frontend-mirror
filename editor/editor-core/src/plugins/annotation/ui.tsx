import React from 'react';
import { ReactNodeView, ForwardRef } from '../../nodeviews';
import WithPluginState from '../../ui/WithPluginState';
import { stateKey as reactPluginKey } from '../../plugins/base/pm-plugins/react-nodeview';

import { InlineCommentPluginState } from './types';
import { inlineCommentPluginKey } from './pm-plugins/plugin-factory';

export class AnnotationNodeView extends ReactNodeView {
  createDomRef() {
    return document.createElement('span');
  }

  getContentDOM() {
    const dom = document.createElement('span');
    dom.className = 'fabric-editor-annotation';
    return { dom };
  }

  render(_props: {}, forwardRef: ForwardRef) {
    return (
      <WithPluginState
        plugins={{
          inlineCommentState: inlineCommentPluginKey,
          selectionState: reactPluginKey,
        }}
        editorView={this.view}
        render={({
          inlineCommentState,
        }: {
          inlineCommentState: InlineCommentPluginState;
        }) => {
          const id = this.node.attrs.id;
          return (
            <span
              style={{
                backgroundColor:
                  inlineCommentState.annotations[id] === false
                    ? 'rgba(255, 196, 0, 0.4)'
                    : 'transparent',
              }}
              ref={forwardRef}
            />
          );
        }}
      />
    );
  }
}
