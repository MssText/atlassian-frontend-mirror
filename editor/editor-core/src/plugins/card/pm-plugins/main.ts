import { Plugin, EditorState } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import { CardProvider } from '@atlaskit/editor-common/provider-factory';
import { CardPlatform } from '@atlaskit/smart-card';

import { CardPluginState, Request } from '../types';
import reducer from './reducers';
import { setProvider, resolveCard } from './actions';
import { replaceQueuedUrlWithCard } from './doc';
import { PMPluginFactoryParams } from '../../../types';
import { InlineCard } from '../nodeviews/inlineCard';
import { BlockCard } from '../nodeviews/blockCard';
import { EmbedCard } from '../nodeviews/embedCard';
import { pluginKey } from './plugin-key';

export { pluginKey } from './plugin-key';

export const getPluginState = (editorState: EditorState) =>
  pluginKey.getState(editorState) as CardPluginState | undefined;

const handleResolved = (view: EditorView, request: Request) => (
  resolvedCard: any,
) => {
  replaceQueuedUrlWithCard(
    request.url,
    resolvedCard,
    request.analyticsAction,
  )(view.state, view.dispatch);
  return resolvedCard;
};

const handleRejected = (view: EditorView, request: Request) => () => {
  view.dispatch(resolveCard(request.url)(view.state.tr));
};

export type OutstandingRequests = { [key: string]: Promise<string> };

export const resolveWithProvider = (
  view: EditorView,
  outstandingRequests: OutstandingRequests,
  provider: CardProvider,
  request: Request,
) => {
  return (outstandingRequests[request.url] = provider
    .resolve(request.url, request.appearance)
    .then(resolvedCard => {
      delete outstandingRequests[request.url];
      return resolvedCard;
    })
    .then(handleResolved(view, request), handleRejected(view, request)));
};

export const createPlugin = (
  platform: CardPlatform,
  allowResizing?: boolean,
  fullWidthMode?: boolean,
) => ({
  portalProviderAPI,
  eventDispatcher,
  providerFactory,
  dispatchAnalyticsEvent,
}: PMPluginFactoryParams) =>
  new Plugin({
    state: {
      init(): CardPluginState {
        return {
          requests: [],
          provider: null,
          cards: [],
          showLinkingToolbar: false,
        };
      },

      apply(tr, pluginState: CardPluginState) {
        // update all the positions
        const remappedState = {
          ...pluginState,
          requests: pluginState.requests.map(request => ({
            ...request,
            pos: tr.mapping.map(request.pos),
          })),

          cards: pluginState.cards.map(card => ({
            ...card,
            pos: tr.mapping.map(card.pos),
          })),
        };

        // apply any actions
        const meta = tr.getMeta(pluginKey);
        if (meta) {
          const nextPluginState = reducer(remappedState, meta);
          return nextPluginState;
        }

        return remappedState;
      },
    },

    view(view: EditorView) {
      // listen for card provider changes
      const handleProvider = (
        _: 'cardProvider',
        provider?: Promise<CardProvider>,
      ) => {
        if (!provider) {
          return;
        }

        provider.then((cardProvider: CardProvider) => {
          const { state, dispatch } = view;
          dispatch(setProvider(cardProvider)(state.tr));
        });
      };

      providerFactory.subscribe('cardProvider', handleProvider);
      const outstandingRequests: OutstandingRequests = {};

      return {
        update(view: EditorView, prevState: EditorState) {
          const currentState = getPluginState(view.state);
          const oldState = getPluginState(prevState);

          if (currentState && currentState.provider) {
            // find requests in this state that weren't in the old one
            const newRequests = oldState
              ? currentState.requests.filter(
                  req =>
                    !oldState.requests.find(
                      oldReq =>
                        oldReq.url === req.url && oldReq.pos === req.pos,
                    ),
                )
              : currentState.requests;

            // ask the CardProvider to resolve all new requests
            const { provider } = currentState;
            newRequests.forEach(request => {
              resolveWithProvider(view, outstandingRequests, provider, request);
            });
          }
        },

        destroy() {
          // cancel all outstanding requests
          Object.keys(outstandingRequests).forEach(url =>
            Promise.reject(outstandingRequests[url]),
          );

          providerFactory.unsubscribe('cardProvider', handleProvider);
        },
      };
    },

    props: {
      nodeViews: {
        inlineCard: (node, view, getPos) => {
          return new InlineCard(
            node,
            view,
            getPos,
            portalProviderAPI,
            eventDispatcher,
            {
              providerFactory,
            },
            undefined,
            true,
          ).init();
        },
        blockCard: (node, view, getPos) => {
          return new BlockCard(
            node,
            view,
            getPos,
            portalProviderAPI,
            eventDispatcher,
            {
              providerFactory,
              platform,
            },
            undefined,
            true,
          ).init();
        },
        embedCard: (node, view, getPos) => {
          return new EmbedCard(
            node,
            view,
            getPos,
            portalProviderAPI,
            eventDispatcher,
            {
              eventDispatcher,
              providerFactory,
              platform,
              allowResizing,
              fullWidthMode: fullWidthMode,
              dispatchAnalyticsEvent,
            },
            undefined,
            true,
          ).init();
        },
      },
    },

    key: pluginKey,
  });
