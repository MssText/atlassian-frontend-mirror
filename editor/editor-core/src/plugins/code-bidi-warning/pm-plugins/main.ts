import { Plugin } from 'prosemirror-state';
import { codeBidiWarningMessages } from '@atlaskit/editor-common/messages';

import { PMPluginFactoryParams } from '../../../types';

import { codeBidiWarningPluginKey } from '../plugin-key';

import {
  createBidiWarningsDecorationSetFromDoc,
  createPluginState,
  getPluginState,
} from './plugin-factory';

export const createPlugin = ({
  dispatch,
  reactContext,
}: PMPluginFactoryParams) => {
  const intl = reactContext().intl;

  const codeBidiWarningLabel = intl.formatMessage(
    codeBidiWarningMessages.label,
  );

  return new Plugin({
    key: codeBidiWarningPluginKey,
    state: createPluginState(dispatch, (state) => {
      return {
        decorationSet: createBidiWarningsDecorationSetFromDoc({
          doc: state.doc,
          codeBidiWarningLabel,
        }),
        codeBidiWarningLabel,
      };
    }),
    props: {
      decorations: (state) => {
        const { decorationSet } = getPluginState(state);

        return decorationSet;
      },
    },
  });
};
