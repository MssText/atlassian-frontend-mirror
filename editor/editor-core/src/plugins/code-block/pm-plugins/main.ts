import { EditorState, Plugin } from 'prosemirror-state';
import { codeBlockNodeView } from '../nodeviews/code-block';
import { CommandDispatch, PMPluginFactoryParams } from '../../../types';
import { pluginKey } from '../plugin-key';
import { findCodeBlock } from '../utils';

export type CodeBlockState = {
  pos: number | null;
};

export const getPluginState = (state: EditorState): CodeBlockState =>
  pluginKey.getState(state);

export const setPluginState = (stateProps: Object) => (
  state: EditorState,
  dispatch: CommandDispatch,
): boolean => {
  const pluginState = getPluginState(state);
  dispatch(
    state.tr.setMeta(pluginKey, {
      ...pluginState,
      ...stateProps,
    }),
  );
  return true;
};

export type CodeBlockStateSubscriber = (state: CodeBlockState) => any;

export const createPlugin = ({ dispatch }: PMPluginFactoryParams) =>
  new Plugin({
    state: {
      init(_, state): CodeBlockState {
        const node = findCodeBlock(state, state.selection);
        return {
          pos: node ? node.pos : null,
        };
      },
      apply(
        tr,
        pluginState: CodeBlockState,
        _oldState,
        newState,
      ): CodeBlockState {
        if (tr.docChanged || tr.selectionSet) {
          const { selection } = newState;

          const node = findCodeBlock(newState, selection);
          const newPluginState: CodeBlockState = {
            ...pluginState,
            pos: node ? node.pos : null,
          };
          return newPluginState;
        }
        return pluginState;
      },
    },
    key: pluginKey,
    props: {
      nodeViews: {
        codeBlock: codeBlockNodeView(),
      },
    },
  });
