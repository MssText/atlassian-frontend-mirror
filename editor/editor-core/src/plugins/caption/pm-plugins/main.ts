import {
  EditorState,
  Plugin,
  TextSelection,
  Transaction,
} from 'prosemirror-state';
import { pluginKey } from './plugin-key';
import captionNodeView from './../nodeviews';
import { PortalProviderAPI } from '../../../ui/PortalProvider';
import { EventDispatcher, Dispatch } from '../../../event-dispatcher';
import { ProviderFactory } from '@atlaskit/editor-common';

export default (
  portalProviderAPI: PortalProviderAPI,
  eventDispatcher: EventDispatcher,
  providerFactory: ProviderFactory,
  dispatch: Dispatch,
) =>
  new Plugin({
    appendTransaction(
      _transactions: Transaction[],
      oldState: EditorState,
      newState: EditorState,
    ): Transaction | void {
      if (
        !newState.selection.eq(oldState.selection) &&
        oldState.selection instanceof TextSelection &&
        oldState.selection.empty
      ) {
        const oldSelectedNode = oldState.doc.nodeAt(
          oldState.selection.from - 1,
        );

        if (
          oldSelectedNode &&
          oldSelectedNode.type === oldState.schema.nodes.caption &&
          oldSelectedNode.childCount === 0
        ) {
          const { tr } = newState;
          tr.delete(oldState.selection.from - 1, oldState.selection.from);
          tr.setMeta('scrollIntoView', false);
          return tr;
        }
      }
    },
    key: pluginKey,
    props: {
      nodeViews: {
        caption: captionNodeView(portalProviderAPI, eventDispatcher),
      },
    },
  });
