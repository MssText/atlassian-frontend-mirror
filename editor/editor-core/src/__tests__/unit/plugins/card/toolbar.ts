import { ProviderFactory } from '@atlaskit/editor-common';
import { IntlProvider } from 'react-intl';
import createEditorFactory from '@atlaskit/editor-test-helpers/create-editor';
import {
  doc,
  p,
  inlineCard,
  blockCard,
} from '@atlaskit/editor-test-helpers/schema-builder';
import {
  floatingToolbar,
  messages as floatingToolbarMessages,
} from '../../../../plugins/card/toolbar';
import { pluginKey } from '../../../../plugins/card/pm-plugins/main';

import commonMessages, {
  linkToolbarMessages,
  linkMessages,
} from '../../../../messages';

import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import OpenIcon from '@atlaskit/icon/glyph/shortcut';
import UnlinkIcon from '@atlaskit/icon/glyph/editor/unlink';

import {
  FloatingToolbarButton,
  FloatingToolbarDropdown,
} from '../../../../plugins/floating-toolbar/types';
import { Command } from '../../../../types';
import { getToolbarItems } from '../floating-toolbar/_helpers';
import { DropdownOptionT } from '../../../../plugins/floating-toolbar/ui/types';

describe('card', () => {
  const createEditor = createEditorFactory();
  const providerFactory = new ProviderFactory();

  const editor = (doc: any) => {
    return createEditor({
      doc,
      providerFactory,
      editorProps: {
        UNSAFE_cards: {},
      },
      pluginKey,
    });
  };

  describe('toolbar', () => {
    const intlProvider = new IntlProvider({ locale: 'en' });
    const { intl } = intlProvider.getChildContext();

    const visitTitle = intl.formatMessage(linkMessages.openLink);
    const unlinkTitle = intl.formatMessage(linkToolbarMessages.unlink);
    const removeTitle = intl.formatMessage(commonMessages.remove);

    it('has an unlink button for inlineCard', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const unlinkButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === unlinkTitle,
      );

      expect(unlinkButton).toBeDefined();
      expect(unlinkButton).toMatchObject({
        icon: UnlinkIcon,
      });
    });

    it('has an remove button for blockCard', () => {
      const { editorView } = editor(
        doc(
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      );

      expect(removeButton).toBeDefined();
      expect(removeButton).toMatchObject({
        appearance: 'danger',
        icon: RemoveIcon,
      });
    });

    it('has a visit button', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(toolbar).toBeDefined();

      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      );

      expect(visitButton).toBeDefined();
      expect(visitButton).toMatchObject({
        icon: OpenIcon,
      });
    });

    it('opens the url in a new window defined on an inline card', () => {
      // @ts-ignore
      global.open = jest.fn();

      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );

      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      ) as FloatingToolbarButton<Command>;

      visitButton.onClick(editorView.state, editorView.dispatch);
      expect(open).toBeCalledWith('http://www.atlassian.com/');
    });

    it('opens the url in a new window via data on an inline card', () => {
      // @ts-ignore
      global.open = jest.fn();

      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              data: {
                url: 'http://www.atlassian.com/',
              },
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const visitButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === visitTitle,
      ) as FloatingToolbarButton<Command>;

      visitButton.onClick(editorView.state, editorView.dispatch);
      expect(open).toBeCalledWith('http://www.atlassian.com/');
    });

    it('deletes a block card', () => {
      const { editorView } = editor(
        doc(
          p('ab'),
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
          p('cd'),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const removeButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === removeTitle,
      ) as FloatingToolbarButton<Command>;

      removeButton.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(doc(p('ab'), p('cd')));
    });

    it('deletes an inline card', () => {
      const { editorView } = editor(
        doc(
          p(
            'ab',
            '{<node>}',
            inlineCard({
              data: {
                title: 'Welcome to Atlassian!',
                url: 'http://www.atlassian.com/',
              },
            })(),
            'cd',
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      const unlinkButton = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'button' && item.title === unlinkTitle,
      ) as FloatingToolbarButton<Command>;

      unlinkButton.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(
        doc(p('abWelcome to Atlassian!cd')),
      );
    });

    it('has no toolbar items when url via url attr is invalid', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'javascript:alert(document.domain)',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(getToolbarItems(toolbar!, editorView).length).toEqual(0);
    });

    it('has no toolbar items when url via data attr is invalid', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'javascript:alert(document.domain)',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({})(
        editorView.state,
        intl,
        providerFactory,
      );
      expect(getToolbarItems(toolbar!, editorView).length).toEqual(0);
    });

    it('it has view switcher when block cards are enabled', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({ allowBlockCards: true })(
        editorView.state,
        intl,
        providerFactory,
      );
      const dropdownTitle = intl.formatMessage(floatingToolbarMessages.inline);

      const dropdown = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'dropdown' && item.title === dropdownTitle,
      );

      expect(toolbar).toBeDefined();
      expect(dropdown).toBeDefined();
      expect(
        (dropdown as FloatingToolbarDropdown<Command>).options,
      ).toHaveLength(2);
    });

    it('it updates the dropdown title for different apperance', () => {
      const { editorView } = editor(
        doc(
          p('ab'),
          '{<node>}',
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
          p('cd'),
        ),
      );

      const toolbar = floatingToolbar({ allowBlockCards: true })(
        editorView.state,
        intl,
        providerFactory,
      );
      const dropdownBlockTitle = intl.formatMessage(
        floatingToolbarMessages.block,
      );

      const dropdown = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'dropdown' && item.title === dropdownBlockTitle,
      );

      expect(dropdown).toBeDefined();
    });

    it('it switches apperance on dropdown option click', () => {
      const { editorView } = editor(
        doc(
          p(
            '{<node>}',
            inlineCard({
              url: 'http://www.atlassian.com/',
            })(),
          ),
        ),
      );

      const toolbar = floatingToolbar({ allowBlockCards: true })(
        editorView.state,
        intl,
        providerFactory,
      );
      const dropdownTitle = intl.formatMessage(floatingToolbarMessages.inline);
      const blockCardTitle = intl.formatMessage(floatingToolbarMessages.block);

      const dropdown = getToolbarItems(toolbar!, editorView).find(
        item => item.type === 'dropdown' && item.title === dropdownTitle,
      ) as FloatingToolbarDropdown<Command> | undefined;

      expect(dropdown).toBeDefined();
      const blockCardOption = (dropdown!.options as Array<
        DropdownOptionT<Command>
      >).filter(option => option.title === blockCardTitle)[0];

      blockCardOption.onClick(editorView.state, editorView.dispatch);
      expect(editorView.state.doc).toEqualDocument(
        doc(
          blockCard({
            url: 'http://www.atlassian.com/',
          })(),
        ),
      );
    });
  });
});
