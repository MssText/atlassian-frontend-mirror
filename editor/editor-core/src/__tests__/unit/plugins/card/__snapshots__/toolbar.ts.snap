// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card toolbar displays toolbar items in correct order for blockCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "id": "editor.link.edit",
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "testId": "link-toolbar-edit-link-button",
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "id": "editor.link.openLink",
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.link.delete",
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`card toolbar displays toolbar items in correct order for embedCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.alignLeft",
    "onClick": [Function],
    "selected": false,
    "title": "Align left",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.alignCenter",
    "onClick": [Function],
    "selected": true,
    "title": "Align center",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.alignRight",
    "onClick": [Function],
    "selected": false,
    "title": "Align right",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.wrapLeft",
    "onClick": [Function],
    "selected": false,
    "title": "Wrap left",
    "type": "button",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.wrapRight",
    "onClick": [Function],
    "selected": false,
    "title": "Wrap right",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "id": "editor.link.edit",
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "testId": "link-toolbar-edit-link-button",
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "id": "editor.link.openLink",
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.link.delete",
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`card toolbar displays toolbar items in correct order for embedCard inside an expand 1`] = `
Array [
  Object {
    "id": "editor.link.edit",
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "testId": "link-toolbar-edit-link-button",
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "id": "editor.link.openLink",
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.link.delete",
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`card toolbar displays toolbar items in correct order for inlineCard 1`] = `
Array [
  Object {
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "id": "editor.link.edit",
    "onClick": [Function],
    "selected": false,
    "showTitle": true,
    "testId": "link-toolbar-edit-link-button",
    "title": "Edit link",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "className": "hyperlink-open-link",
    "icon": [Function],
    "id": "editor.link.openLink",
    "onClick": [Function],
    "title": "Open link in a new tab",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "id": "editor.link.unlink",
    "onClick": [Function],
    "title": "Unlink",
    "type": "button",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.link.delete",
    "onClick": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "title": "Remove",
    "type": "button",
  },
]
`;
